src/he/msg_handlers.c

```diff
@@ -39,28 +39,47 @@ he_return_code_t he_handle_msg_ping(he_conn_t *conn, uint8_t *packet, int length
   }
 
   // We only want to do stuff with pings when the connection is ONLINE
-
   if(conn->state != HE_STATE_ONLINE) {
     return HE_ERR_INVALID_CONN_STATE;
   }
 
-  // Create the pong here
-  he_msg_pong_t response = {0};
-  response.msg_header.msgid = HE_MSGID_PONG;
+  // Return if it's not a valid ping message
+  if(length < sizeof(he_msg_ping_t)) {
+    return HE_ERR_PACKET_TOO_SMALL;
+  }
 
-  // Send pong
-  he_internal_send_message(conn, (uint8_t *)&response, sizeof(he_msg_pong_t));
+  // Get the ping message
+  he_msg_ping_t *ping = (he_msg_ping_t *)packet;
 
-  return HE_SUCCESS;
+  // Create the pong message
+  he_msg_pong_t pong = {0};
+  pong.msg_header.msgid = HE_MSGID_PONG;
+  pong.id = ping->id;
+
+  // Send pong
+  return he_internal_send_message(conn, (uint8_t *)&pong, sizeof(he_msg_pong_t));
 }
 
 he_return_code_t he_handle_msg_pong(he_conn_t *conn, uint8_t *packet, int length) {
   if(conn == NULL || packet == NULL) {
     return HE_ERR_NULL_POINTER;
   }
 
+  // Ignore the message if it's not a valid pong message
+  if(length < sizeof(he_msg_pong_t)) {
+    return HE_SUCCESS;
+  }
+
+  // Get the pong message, and ignore it if the id doesn't match
+  he_msg_pong_t *pong = (he_msg_pong_t *)packet;
+  uint16_t id = ntohs(pong->id);
+  if(id != conn->ping_pending_id) {
+    return HE_SUCCESS;
+  }
+
```

Comment:
Do we need to reset the `ping_pending_id`  after we receive one successful pong message ?

Reply:
Good questions. It's intentional that `pong` message handlers shouldn't reset the `ping_pending_id`. Because:

Each new ping message will reset the `ping_pending_id` anyway. It good to have a single owner of this responsibility to avoid race conditions. 

For example, if we also reset the ping_pending_id in pong message handler, consider this scenario:

1. client sent a ping message: ping_pending_id is set to 1
2. timer is up, the client is trying to send another ping message which has id 2
3. at the same time, client receives the pong message with id 1

both step 2 & step 3 will try to reset ping_pending_id and the behaviour will be nondeterministic. We could use locks to synchronize it, but I don't see the value of having that complexity yet.

Another (more theoretical) case is, when aggressive mode is enabled, it's possible to receive multiple pong messages with same id, and it's expected to raise 3 PONG RECEIVED events according to current design.

Reply:
I think our code is designed for single threaded applications, that's why we don't have any mutex locks.

If that is the case, the steps 2 and 3 cannot happen at same time and It will be sequential. 

If 2 happens first,  `ping_pending` will be updated by client say `3` and then pong message with ` 2` will be received and ignored

If 3 happens first, pong message handler will clear` ping_pending` to INVALID id, and then client will update the ping_pending to 3 and wait for it.


I was suggesting to reset the ID to handle aggressive mode only. If the pending id matches the incoming pong message id, we are raising the event `HE_EVENT_PONG`. In case of aggressive mode, we may be raising three `HE_EVENT_PONG` events for one `he_conn_send_keepalive()` call from client.

Reply:
Discussed with @expressvpn-mariappan-r offline. The case for aggressive mode is a pre-existing issue, we'll handle it separated. There's no other cases supporting resetting `ping_pending_id` in both places, so we'll not do it here.